### PR TITLE
Update canon tags 2

### LINKED ISSUE #2059: CanonRFLensType values are wrong in canonmn_int.cpp
##### **Describe the bug**
`canonRFLensType` values are wrong in https://github.com/Exiv2/exiv2/blob/main/src/canonmn_int.cpp#L2462-L2482 (`257` for all Canon RF lenses, while it should be a different number for each lens except Canon RF 50mm F1.2L USM). 

##### **Expected behavior**
The correct values are described here: https://exiftool.org/TagNames/Canon.html (search "RFLensType").

### PR DESCRIPTION
**Pull request by clanmills**
This super-cedes #1628.  Test 981 is failing because the output from Exif.CanonAF2Id tags have the wrong type/value.

### CODE REVIEW

File: tests/bugfixes/github/test_issue_981.py
Comment (piponazo) [inline tests/bugfixes/github/test_issue_981.py]: 👀 how is possible that this was working previously?

**Comment by piponazo** [comment]
Great to see the tests finally passing! Good team work 💪  @alexvanderberkel @clanmills .

Is the work in this branch finished? Or do you have plans to keep working here?

**Review by piponazo — DISMISSED** [review]

File: src/canonmn_int.cpp
Comment (hassec) [inline src/canonmn_int.cpp:2529]: this closes #2059, just mentioning it so we don't forget

File: tests/bugfixes/github/test_issue_20.py
Comment (hassec) [inline tests/bugfixes/github/test_issue_20.py:29]: Serialnumber of `0` seems weird. Is this info not included in all files? 
Are there other tests where we can see if this decoding works?

After going through the entire PR, I don't see a single value that isn't zero... :thinking:

File: tests/bugfixes/github/test_issue_981.py
Comment (hassec) [inline tests/bugfixes/github/test_issue_981.py:16]: Why this change? Now we do `exiv2 -pv` twice for `filenameC` and `exiv2 -pt` twice for `filenameD` instead of one of these for each filename

File: tests/bugfixes/github/test_issue_981.py
Comment (hassec) [inline tests/bugfixes/github/test_issue_981.py:61]: Did anyone check if this change makes sense? (e.g. validate against exiftool)

**Review by hassec — CHANGES_REQUESTED** [review]
Just a few small points right now.

File: src/tiffimage_int.cpp
Comment (hassec) [inline src/tiffimage_int.cpp:1692]: Why is this one commented out? I think we should add a comment explaining the reason, so we don't forget.

Comment (clanmills) [reply src/tiffimage_int.cpp:1692]: I don't know.  @alexvanderberkel did that.  My change to that line was cosmetic (aligning everything).

File: tests/bugfixes/github/test_issue_20.py
Comment (clanmills) [reply tests/bugfixes/github/test_issue_20.py:29]: @alexvanderberkel did that.  Alex?  That is a read/write field.  You can test it with ExifTool.  So, you can probably modify one of the existing test images to add a test for this.

```
734 rmills@rmillsm1:~/gnu/github/exiv2/main/test/data $ exiv2 -K Exif.Image.Make *.jpg 2>/dev/null | grep Canon 
FurnaceCreekInn.jpg   Exif.Image.Make                              Ascii       6  Canon
FurnaceCreekInn2.jpg  Exif.Image.Make                              Ascii       6  Canon
FurnaceCreekInn3.jpg  Exif.Image.Make                              Ascii       6  Canon
exiv2-bug1202.jpg     Exif.Image.Make                              Ascii       6  Canon
exiv2-bug1231a.jpg    Exif.Image.Make                              Ascii       6  Canon
exiv2-bug1231b.jpg    Exif.Image.Make                              Ascii       6  Canon
exiv2-bug445.jpg      Exif.Image.Make                              Ascii       6  Canon
exiv2-bug501.jpg      Exif.Image.Make                              Ascii       6  Canon
exiv2-bug528.jpg      Exif.Image.Make                              Ascii       6  Canon
exiv2-bug876.jpg      Exif.Image.Make                              Ascii       6  Canon
exiv2-canon-eos-20d.jpg  Exif.Image.Make                              Ascii       6  Canon
exiv2-canon-eos-300d.jpg  Exif.Image.Make                              Ascii       6  Canon
exiv2-canon-eos-d30.jpg  Exif.Image.Make                              Ascii       6  Canon
exiv2-canon-powershot-a520.jpg  Exif.Image.Make                              Ascii       6  Canon
exiv2-canon-powershot-s40.jpg  Exif.Image.Make                              Ascii       6  Canon
issue_1912_poc.jpg    Exif.Image.Make                              Ascii       6  Canon
smiley2.jpg           Exif.Image.Make                              Ascii       6  Canon
735 rmills@rmillsm1:~/gnu/github/exiv2/main/test/data $ 
```

File: tests/bugfixes/github/test_issue_981.py
Comment (clanmills) [reply tests/bugfixes/github/test_issue_981.py:16]: I think I made an error when I invented the test in 2019. So, I fixed it.  Files C and D are concerned with NumPoints and Unusbles.  I want to test both the metadata (-pv) and pretty printer (-pt).

Comment (clanmills) [reply tests/bugfixes/github/test_issue_981.py:16]: Crap you're right.  I've got the change wrong.  I'll fix that.

File: tests/bugfixes/github/test_issue_20.py
Comment (alexvanderberkel) [reply tests/bugfixes/github/test_issue_20.py:29]: will do that

Comment (clanmills) [reply tests/bugfixes/github/test_issue_20.py:29]: @hassec
```
879 rmills@rmillsmm-local:/Users/Shared/Jenkins/Home/userContent/testfiles/ExifTool/Canon $ exiftool CanonEOS_R6.jpg | grep -i serial
Internal Serial Number          : KQ0008148
Serial Number                   : 012021000024
Lens Serial Number              : 8003003431
880 rmills@rmillsmm-local:/Users/Shared/Jenkins/Home/userContent/testfiles/ExifTool/Canon $ exiv2 -K Exif.CanonLe.LensSerialNumber CanonEOS_R6.jpg 
Exif.CanonLe.LensSerialNumber                SLong       1  872416128
881 rmills@rmillsmm-local:/Users/Shared/Jenkins/Home/userContent/testfiles/ExifTool/Canon $ 
```

I've looked at the ExifTool test files and I believe this is working.  I documented how to obtain them: https://exiv2.org/book/index.html#8-6

**BUT BUT BUT _Wait a minute_ **
I extracted the .exv (with a view to putting it in the test suite) and it's not correct.

```
880 rmills@rmillsmm-local:/Users/Shared/Jenkins/Home/userContent/testfiles/ExifTool/Canon $ exiv2 -K Exif.CanonLe.LensSerialNumber CanonEOS_R6.jpg 
Exif.CanonLe.LensSerialNumber                SLong       1  872416128
881 rmills@rmillsmm-local:/Users/Shared/Jenkins/Home/userContent/testfiles/ExifTool/Canon $ exiv2 -ea --verbose CanonEOS_R6.jpg 
File 1/1: CanonEOS_R6.jpg
Writing Exif data from CanonEOS_R6.jpg to ./CanonEOS_R6.exv
Writing XMP data from CanonEOS_R6.jpg to ./CanonEOS_R6.exv
882 rmills@rmillsmm-local:/Users/Shared/Jenkins/Home/userContent/testfiles/ExifTool/Canon $ exiv2 -K Exif.CanonLe.LensSerialNumber CanonEOS_R6.exv
Exif.CanonLe.LensSerialNumber                SLong       1  28
883 rmills@rmillsmm-local:/Users/Shared/Jenkins/Home/userContent/testfiles/ExifTool/Canon $ 
```
If this PR is merged without solving this, we should open an issue concerning the .exv.

_**Other points**_

1.  We don't have a bug report for those canon tag changes. 
2.  We should investigate all the canon tag changes using the ExifTool test suite and compare our results with ExifTool.

**Comment by clanmills** [comment]
The code is correctly reads LensSerialNumber on the attached  ExifTool test file.
```
935 rmills@rmillsmm-local:~/temp $ ls -l CanonEOS_R6.jpg 
-rw-r--r--+ 1 rmills  staff  55545 11 Feb 13:23 CanonEOS_R6.jpg
936 rmills@rmillsmm-local:~/temp $ exiftool CanonEOS_R6.jpg  | grep Lens | grep Serial
Lens Serial Number              : 8003003431
937 rmills@rmillsmm-local:~/temp $ exiv2 -K Exif.Photo.LensSerialNumber CanonEOS_R6.jpg 
Exif.Photo.LensSerialNumber                  Ascii      11  8003003431
938 rmills@rmillsmm-local:~/temp $ 
```
The code also successfully modifies LensSerialNumber
```
938 rmills@rmillsmm-local:~/temp $ cp CanonEOS_R6.jpg  foo.jpg
939 rmills@rmillsmm-local:~/temp $ exiv2 -M'set Exif.Photo.LensSerialNumber 1343003008' foo.jpg 
940 rmills@rmillsmm-local:~/temp $ exiv2 -K Exif.Photo.LensSerialNumber foo.jpg 
Exif.Photo.LensSerialNumber                  Ascii      11  1343003008
```

The code also successfully creates a .EXV file which can be read:
```
941 rmills@rmillsmm-local:~/temp $ exiv2 -ee --force foo.jpg
942 rmills@rmillsmm-local:~/temp $ exiv2 -K Exif.Photo.LensSerialNumber foo.exv 
Exif.Photo.LensSerialNumber                  Ascii      11  1343003008
943 rmills@rmillsmm-local:~/temp $ 
```

File: tests/bugfixes/github/test_issue_981.py
Comment (clanmills) [reply tests/bugfixes/github/test_issue_981.py]: @hassec Are you talking about every change relating to Canon tags, or specifically the AutoFocus changes?   I can't say anything about all the changes relating to Canon tags.  The AutoFocus tags were carefully researched and added to Exiv2 in August 2019 with a lot of input from @derselbst.  The code in this PR passes the tests which we added in 2019.  You'll have to look at #981 for details concerning the Canon/AF support in exiv2 and and exiftool.

**Comment by alexvanderberkel** [comment]
I think we are actually done with the PR or have I missed anything?

File: tests/bugfixes/github/test_issue_981.py
Comment (clanmills) [reply tests/bugfixes/github/test_issue_981.py]: Yes.  The Canon.AFxxx behaviour is unchanged since 0.27.3.  I modified the test slightly because a couple of new fields were add `Exif.Canon.AFMicroAdj` and `Exif.Canon.AFConfig`.  I don't know the genesis of these fields.  @piponazo @alexvanderberkel @hassec and @neheb get a mention in the blame listing:

**AFConfig**
```bash
534 rmills@rmillsm1:~/gnu/github/exiv2/main $ git blame src/canonmn_int.cpp | grep AFConfig 
bdd8a386b src/canonmn_int.cpp (Christoph Hasse    2021-06-08 21:56:04 +0200  637)         {0x4028, "AFConfig", N_("AFConfig"), N_("AFConfig"), canonId, makerTags, unsignedShort, -1, printValue},
ff2ffb190 src/canonmn_int.cpp (Alex Esseling      2021-01-12 20:55:29 +0100 1230)     //Canon AFConfig Tags
6da49fd29 src/canonmn_int.cpp (Rosen Penev        2021-05-17 16:54:53 -0700 1232)          {0x0001, "AFConfigTool", N_("AF Config Tool"), N_("AF Config Tool"), canonAfCId, makerTags, signedLong, -1, printValue},
```

**AFMicroAdj**
```
535 rmills@rmillsm1:~/gnu/github/exiv2/main $ git blame src/canonmn_int.cpp | grep AFMicroAdj
bdd8a386b src/canonmn_int.cpp (Christoph Hasse    2021-06-08 21:56:04 +0200  628)         {0x4013, "AFMicroAdj", N_("AFMicroAdj"), N_("AFMicroAdj"), canonId, makerTags, unsignedShort, -1, printValue},
ff2ffb190 src/canonmn_int.cpp (Alex Esseling      2021-01-12 20:55:29 +0100  906)     // Canon AFMicroAdjMode Quality Info, tag 0x0001
6da49fd29 src/canonmn_int.cpp (Rosen Penev        2021-05-17 16:54:53 -0700  907)     constexpr TagDetails canonAFMicroAdjMode[] = {
bdd8a386b src/canonmn_int.cpp (Christoph Hasse    2021-06-08 21:56:04 +0200  914)      // Canon AFMicroAdj Info Tag
6da49fd29 src/canonmn_int.cpp (Rosen Penev        2021-05-17 16:54:53 -0700  916)          {0x0001, "AFMicroAdjMode", N_("AFMicroAdjMode"), N_("AFMicroAdjMode"), canonAfMiAdjId, makerTags, signedLong, -1, EXV_PRINT_TAG(canonAFMicroAdjMode)},
bdd8a386b src/canonmn_int.cpp (Christoph Hasse    2021-06-08 21:56:04 +0200  917)          {0x0002, "AFMicroAdjValue", N_("AF Micro Adj Value"), N_("AF Micro Adj Value"), canonAfMiAdjId, makerTags, signedRational, -1, printValue},
de036b2b2 src/canonmn_int.cpp (Luis Díaz Más      2022-02-09 23:00:03 +0100  918)          {0xffff, "(UnknownCanonAFMicroAdjTag)", "(UnknownCanonAFMicroAdjTag)", N_("Unknown Canon AFMicroAdj tag"), canonAfMiAdjId, makerTags, signedShort, 1, printValue}
536 rmills@rmillsm1:~/gnu/github/exiv2/main $ 
 ```

File: tests/bugfixes/github/test_issue_20.py
Comment (clanmills) [reply tests/bugfixes/github/test_issue_20.py:29]: @hassec.  I was confused when I thought there was an issue concerning Exif.CanonLe.LensSerialNumber.  I have investigated and provided evidence that it is working correctly.  https://github.com/Exiv2/exiv2/pull/2087#issuecomment-1036322420

**Comment by clanmills** [comment]
Gentlemen:  I believe this is working and can be merged.

@alexvanderberkel.  We should have a bug report to justify those changes.  I believe lots of new tags have been introduced and it would be good to have a bug report that says "Please provide support for the following Canon MakerNote Tags" and provide a list.

Before merging and closing this PR, it would be good to see evidence that:
1.  The Tags listed in the bug report have been tested (using the Exiv2 Canon Test Images)
- ideally, test should be added the exiv2 test suite for the new tags.
- if this requires adding lots of "fat" images to the test suite, terminal transcript evidence in this PR is sufficient.
2. Compare exiv2 v ExifTool for the tags listed in the bug report.

@1div0.  Can you help @alex with this?  Obtaining the ExifTool test images is documented in my book:  https://exiv2.org/book/index.html#8-6

**Comment by piponazo** [comment]
> Gentlemen: I believe this is working and can be merged.

However many of the CI jobs are failing. Until all the relevant CI jobs are passing we should not merge this PR.

**Comment by clanmills** [comment]
Crap.  I don't see that.  The last time I looked, they were green.

@alexvanderberkel is kind of busy.  This PR has been hanging about for 6 months.  How are we going to get this done?

**Comment by clanmills** [comment]
I'll investigate this CI failures.  @alexvanderberkel Can you please submit a bug report enumerating the keys to be supported.  When the CI is green, perhaps @1div0 and/or @sridharb1 can contribute to getting this tested.

**Comment by hassec** [comment]
it's just the new tests that are failing because they were created without the new output of this MR.    

But the test automatically creates new reference files with the changes and stores them in a tmp folder. This is reported in the ouput: 
```
Error:  The output of the testcase mismatch the reference
[INFO] The output has been saved to file D:\a\exiv2\exiv2\test\tmp\CanonEF100mmF2.8LMacroISUSM.exv.out
[INFO] simply_diff:
D:\a\exiv2\exiv2\test/data\test_reference_files\CanonEF100mmF2.8LMacroISUSM.exv.out: 173 lines
D:\a\exiv2\exiv2\test\tmp\CanonEF100mmF2.8LMacroISUSM.exv.out: 180 lines
```

So I'd suggest that someone should just run the test locally, copy the new references from `exiv2/test/tmp` over into `exiv2/test/test_reference_files/` and commit and push them. 

Then we can all have a look together if the diff makes sense and then hopefully merge :tada:

**Comment by clanmills** [comment]
Thanks @hassec for the explanation.  And Thanks @postscript-dev for your web/documentation changes.

Q.  How did regressionTest get into this PR?  It hadn't been invented when we were working on this PR last week.
Q.  Has regressionTest been documented in README.md when test types _bash, bugfix, unit, lens, tiff and version_ are tested? 

It is regression that's failing on macOS (and presumably everywhere else).  I'll fix and update th PR.

**Comment by clanmills** [comment]
It must be Christmas.  All the lights have turned **Green!**.

File: src/canonmn_int.cpp
Comment (clanmills) [reply src/canonmn_int.cpp:2477]: @hassec I believe this is included in the PR.  I've linked #2059 to this PR so it closed when this is merged.

```bash
581 rmills@rmillsm1:~/gnu/github/exiv2/main/build $ grep -n 'Canon RF 50mm F1.8 STM' ../src/*.cpp 
../src/canonmn_int.cpp:2019:        {61182, "Canon RF 50mm F1.8 STM"                                    }, // 8
../src/canonmn_int.cpp:2477:        { 280, N_("Canon RF 50mm F1.8 STM")  }
582 rmills@rmillsm1:~/gnu/github/exiv2/main/build $ 
```

Do  you know why it's in two places, @postscript-dev ?  It's localised in one `N_("....")` and not in the other.

**Comment by postscript-dev** [comment]
@clanmills: For some reason I couldn't reply to the review above, so have done so here.

> @hassec I believe this is included in the PR. I've linked https://github.com/Exiv2/exiv2/issues/2059 to this PR so it closed when this is merged.
```
581 rmills@rmillsm1:~/gnu/github/exiv2/main/build $ grep -n 'Canon RF 50mm F1.8 STM' ../src/*.cpp 
../src/canonmn_int.cpp:2019:        {61182, "Canon RF 50mm F1.8 STM"                                    }, // 8
../src/canonmn_int.cpp:2477:        { 280, N_("Canon RF 50mm F1.8 STM")  }
```
> Do you know why it's in two places, @postscript-dev ? It's localised in one N_("....") and not in the other.

I have only started using Canon files recently but I see that Canon adds more than one lens related tag. For example, in [this](https://user-images.githubusercontent.com/97848353/150316985-16cdf438-4e78-4379-a3ee-df27aeceb585.jpg) file from #2057, there are:
```
$ exiv2 -g lens/i Canon_RF_24-105_F4-7.1_IS_STM.jpg
Exif.CanonCs.LensType                        Short       1  Unknown Lens (61182)
Exif.CanonCs.Lens                            Short       3  24.0 - 105.0 mm
Exif.CanonFi.RFLensType                      SShort      1  Canon RF 24-105mm F4-7.1 IS STM
Exif.CanonCf.LensAFStopButton                Short       1  0
Exif.Canon.LensModel                         Ascii     138  RF24-105mm F4-7.1 IS STM
Exif.CanonLe.LensSerialNumber                SLong       1  -1509818368
Exif.CanonAfC.USMLensElectronicMF            SLong       1  Enable After AF
Exif.CanonAfC.LensDriveWhenAFImpossible      SLong       1  Continue Focus Search
Exif.Photo.LensSpecification                 Rational    4  24/1 105/1 0/1 0/1
Exif.Photo.LensModel                         Ascii      25  RF24-105mm F4-7.1 IS STM
Exif.Photo.LensSerialNumber                  Ascii      11  000002a620
``` 
(note: In #2057, I am going to fix the `Exif.CanonCs.LensType` translated output).

The two values that you found in *src/canonmn_int.cpp* are from the `Exif.CanonCs.LensType` and `Exif.CanonFi.RFLensType` respectively. It is possible that these tags may not always be the same value.

I looked at some of the lens lists in other makernotes and I didn't see the `N_("...")` function used. As any string we use `N_("...")` with appears in the Exiv2 localisation project, it might be better to remove them in this case.

**Comment by clanmills** [comment]
Ah right, thanks @postscript-dev.  Lens recognition is horrible.

You're right.  That string should probably not be localised.  Strings pass unchanged if not translated.

I'm happy with your explanation that CanonCs and CanonFi are different families _(and probably mutually exclusive)_.

If I have sufficient permission, I'll mark this "change request blocker" as resolved.

I hope than @alexvanderberkel will provide an issue report to enumerate the tags which are intended to be modified/recognised in this PR.

If there are no further "blockers" and the CI is green I hope somebody will approve and merge this PR.

**Comment by clanmills** [comment]
We still have "1 change requested".  I can't find the outstanding request.

**Review by hassec — APPROVED** [review]

**Pull request merged**
